pkg/autoscaler/metrics/stats_scraper.go

```diff
@@ -181,7 +189,111 @@ func (s *serviceScraper) Scrape(window time.Duration) (Stat, error) {
 	if readyPodsCount == 0 {
 		return emptyStat, nil
 	}
-	return s.scrapeService(window, readyPodsCount)
+
+	startTime := time.Now()
+	scrapeTime := time.Since(startTime)
+	pkgmetrics.RecordBatch(s.statsCtx, scrapeTimeM.M(float64(scrapeTime.Milliseconds())))
+
+	if s.podsAddressable {
+		stat, err := s.scrapePods(readyPodsCount)
+		// Some pods were scraped, but not enough.
+		if err != errNoPodsScraped {
+			return stat, err
+		}
+		// Else fall back to service scrape.
+	}
+	stat, err := s.scrapeService(window, readyPodsCount)
+	if err == nil {
+		// If err == nil, this means that we failed to scrape all pods, but service worked
+		// thus it is probably a mesh case.
+		s.podsAddressable = false
+	}
+	return stat, err
+}
+
+func (s *serviceScraper) scrapePods(readyPods int) (Stat, error) {
+	pods, err := s.podAccessor.PodIPsByAge()
+	if err != nil {
+		return emptyStat, err
+	}
+	// Race condition when scaling to 0, where the check above
+	// for endpoint count worked, but here we had no ready pods.
+	if len(pods) == 0 {
+		return emptyStat, nil
+	}
+
+	frpc := float64(readyPods)
+	sampleSizeF := populationMeanSampleSize(frpc)
+	sampleSize := int(sampleSizeF)
+	results := make(chan Stat, sampleSize)
+
+	grp := errgroup.Group{}
+	idx := int32(-1)
+	for i := 0; i < sampleSize; i++ {
+		grp.Go(func() error {
+			for {
+				// Acquire next pod.
+				myIdx := int(atomic.AddInt32(&idx, 1))
+				// All out?
+				if myIdx >= len(pods) {
+					return errPodsExahusted
+				}
+
+				// Scrape!
+				target := "http://" + pods[myIdx] + ":" + portAndPath
+				stat, err := s.sClient.Scrape(target)
+				if err == nil {
```

Comment:
kind of sad that the real error gets swallowed here.. is there any non-awful way to return one of the actual errors instead of errPodsExhausted - maybe a `chan error` we can write the error to and pull the first error out to log/wrap from the block on L258 or something? Would help debugging if we ever find a system falling back to mesh scraping unexpectedly or failing to scrape

Reply:
There might be more than one, which makes packaging hard.  HttpScraper should log (or perhaps right here, though we have no logging right now).